### ISSUE TITLE
TopicClient fails to send message if generated SAS token has expiration that is far in future

### ISSUE THREAD

**Issue by sergsalo** [Azure/azure-service-bus-dotnet#399]
Neither TopicClient nor the QueueClient are working with long living SAS tokens. I set the token to be valid for a year.  Here is token example:
SharedAccessSignature sr=sb%3A%2F%2Fxxxxxx.servicebus.windows.net%2Ftest&sig=TmyxxxxxxxxVPqmXlb8uKb7iJhk%3D&se=1530579259&skn=RootManageSharedAccessKey

Error: "ArgumentOutOfRangeException: Time-out interval must be less than 2^32-2."
Parameter name: dueTime

nuget "Microsoft.Azure.ServiceBus", version: 2.0.0
It works fine with "WindowsAzure.ServiceBus" nuget package.

**Comment by nemakam** [comment]
This is throwing in `ActiveClientLinkManager.ChangeRenewTimer()`. This is due to system.threading.timer doesn't allow for timer to be scheduled for that long in the future. 
Will work on a fix for this.

**Comment by SeanFeldman** [comment]
@nemakam at the same token, if `System.Threading.Timer` is still going to be used, should `ActiveClientLinkManager` [implement `IDisposable`](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose#implementing-the-dispose-pattern-for-a-derived-class) as well?

**Comment by nemakam** [comment]
The main use of `IDisposable` in a library is to make sure the users of the library follow the disposable pattern the right way and not leak resources after they are done using it. 
For our client, since Connection is a heavy object, we do not want it to be disposable. We do not want to encourage people using it like
`using (var qc = new QueueClient)`
So none of the clients (Queue/Topic/Sub/Sender/Receiver) is going to implement `IDisposable`. 
Given this stance, everything now needs to be bound to `CloseAsync()` of the client and not the `Dispose()` of the classes. And so, I don't see any strong reason why its required to change it to implement IDisposable now. Thoughts?

**Comment by SeanFeldman** [comment]
Clients still could be created in a manner that would cause heavy resource consumption. No one is going to stop the code from instantiating a client in a tight loop for example
```c#
foreach (var something in collection)
{
  await IvokeMethodThatInstantiatesQueueClient(something);
}
```
This was more of a question. The need to invoke `CloseAsync()` could be [eventually addressed in C#](https://github.com/dotnet/roslyn/issues/114).

**Comment by rozele** [comment]
Any updates on this? We have a scenario where we need to have SAS tokens with expiry much farther out than 2^32ms (~49.5days) for `SubscriptionClient`.

**Comment by nemakam** [comment]
@rozele 
#445 will take care of your problem

**Comment by bvandermeide** [comment]
@SeanFeldman & @nemakam,

# TLDR
Please consider a suggested change to `ActiveClientLinkManager.SetRenewCbsTokenTimer` that would capture DateTime.UtcNow and ensures `interval` is at least the value of `TokenRefreshBuffer` before sending into `ChangeRenewTimer`

# Explained

## My Scenario Explained
I ran into this same problem this week using the latest NuGet packages of MassTransit, MassTransit.Azure.ServiceBus.Core, and Microsoft.Azure.ServiceBus.  The issue I ran into was primarily an initialization issue with MassTransit.  (  **_see commented line `sig.TokenTimeToLive = TimeSpan.FromDays(1);`_** )

I believe the issue I was running into is that the MassTransit would check for a null value of TokenTimeToLive and set a default value, but the value is not null and sets it to a 0.00 TimeSpan.  Therefore, the `SetRenewCbsTokenTimer` would get a SAS token that would immediatley expire.  In the process of fixing this bug in my software, I thought there may be a couple adjustments that may make sense to the Microsoft.Azure.ServiceBus code.  Please consider these changes.

````csharp
MassTransit.Bus.Factory.CreateUsingAzureServiceBus(cfg =>
      {
        cfg.UseInMemoryOutbox();
        cfg.UseServiceBusMessageScheduler();

        var host = cfg.Host(new Uri(_connectionResolver.Address), hostCfg =>
        {
          hostCfg.RetryLimit = 3;
          hostCfg.OperationTimeout = TimeSpan.FromMinutes(1);
          hostCfg.SharedAccessSignature(sig =>
                {
                  sig.KeyName = _connectionResolver.KeyName;
                  sig.SharedAccessKey = _connectionResolver.SharedAccessKey;
                  sig.TokenTimeToLive = TimeSpan.FromDays(1);  // without this line MassTransit sets the Token TTL to 0.00 instead of null
                });
        });
        //  ... connect endpoints
      });
````
## ActiveClientLinkManager code today
````csharp

        void SetRenewCbsTokenTimer(ActiveClientLinkObject activeClientLinkObject)
        {
            if (activeClientLinkObject.AuthorizationValidUntilUtc < DateTime.UtcNow)
            {
                return;
            }

            var interval = activeClientLinkObject.AuthorizationValidUntilUtc.Subtract(DateTime.UtcNow) - ActiveClientLinkManager.TokenRefreshBuffer;
            interval = TimeoutHelper.Min(interval, ActiveClientLinkManager.MaxTokenRefreshTime);

            this.ChangeRenewTimer(activeClientLinkObject, interval);
        }

        void ChangeRenewTimer(ActiveClientLinkObject activeClientLinkObject, TimeSpan dueTime)
        {
            if (activeClientLinkObject is ActiveSendReceiveClientLink)
            {
                this.sendReceiveLinkCbsTokenRenewalTimer?.Change(dueTime, Timeout.InfiniteTimeSpan);
            }
            else
            {
                this.requestResponseLinkCbsTokenRenewalTimer?.Change(dueTime, Timeout.InfiniteTimeSpan);
            }
        }
````
## ActiveClientLinkManager.SetRenewCbsTokenTimer change suggestion
````csharp

        void SetRenewCbsTokenTimer(ActiveClientLinkObject activeClientLinkObject)
        {
            var utcNow = DateTime.UtcNow;
            if (activeClientLinkObject.AuthorizationValidUntilUtc < utcNow)
            {
                return;
            }

            var interval = activeClientLinkObject.AuthorizationValidUntilUtc.Subtract(utcNow) - ActiveClientLinkManager.TokenRefreshBuffer;

            if (interval< ActiveClientLinkManager.TokenRefreshBuffer)
                interval = ActiveClientLinkManager.TokenRefreshBuffer;

             interval = TimeoutHelper.Min(interval, ActiveClientLinkManager.MaxTokenRefreshTime);

            this.ChangeRenewTimer(activeClientLinkObject, interval);
        }
````
